File: AoC2018/solutions/20/solution.rs
use std::collections::HashMap;
//Advent of Code 2018 day 20
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::str::Chars;

fn parse_input() -> String{
    let filename = "AoC2018/solutions/20/input";
    let file = File::open(filename).unwrap();
    let reader = BufReader::new(file);
    return reader.lines().next().unwrap().unwrap();
}

fn recurse(mut pos: (i32, i32), directions: & HashMap<char, (i32, i32)>, known: &mut Vec<((i32, i32), char)>, stream: &mut Chars) -> bool{
    loop {
        let c_o = stream.next();
        if c_o.is_none() || c_o.unwrap() == '$'{
            return false;
        }

        let c = c_o.unwrap();
        if directions.contains_key(&c){
            let d = directions.get(&c).unwrap();
            pos.0 +=  d.0;
            pos.1 +=  d.1;
            known.push((*pos, 'x'));
            pos.0 +=  d.0;
            pos.1 +=  d.1;
            known.push((*pos, '.'));
        }
        else {
            match c {
                '(' => print!("("),
                ')' => print!(")"),
                '|' => print!("|"),
                _ => panic!("WTF")
            }
        }

    }




    return true;
}

fn expand_map(regex: String){
    let mut char_iter = regex.chars();
    let mut known = Vec::from([((0, 0), '.')]);
    let directions = HashMap::from([
        ('N', ( 0, -1)),
        ('S', ( 0,  1)),
        ('W', (-1,  0)),
        ('E', ( 1,  0)),
    ]);
    recurse((0, 0), &directions, &mut known, &mut char_iter);
}

pub fn solve() {
    let input = parse_input();
    println!("{}", input);


    let first  = 0;
    let second = 0;
    println!("First:  {}", first);
    println!("Second: {}", second);
}

File: AoC2018/solutions/20/solution.py
#Advent of Code 2018 day 20
from util import *
YEAR = 2018
DAY = 20


def get_data():
    return input_lines(DAY, YEAR)


def generate_map(regex):
    regex = regex[1:-1]
    revealed = {(0, 0)}
    directions = {
        "N": ( 0,  1),
        "S": ( 0, -1),
        "E": ( 1,  0),
        "W": (-1,  0),
    }

    def recurse(start_pos):
        nonlocal revealed
        nonlocal regex
        pos = start_pos

        while len(regex) > 0:
            current_char = regex[0]
            regex = regex[1:]
            if current_char in directions:
                d = directions[current_char]
                pos = pos[0] + d[0], pos[1] + d[1]
                revealed.add(pos)
                pos = pos[0] + d[0], pos[1] + d[1]
                revealed.add(pos)
            else:
                if current_char == "(":
                    recurse(pos)
                elif current_char == "|":
                    pos = start_pos
                else:
                    return

    recurse((0, 0))
    return revealed


def flood(revealed_rooms):
    directions = [(0, 1), (0, -1), (1,  0), (-1, 0)]
    to_check = [((0, 0), 0)]
    checked = set()
    distances = []
    while len(to_check) > 0:
        pos, dist = to_check.pop()
        distances.append(dist)
        checked.add(pos)
        for d in directions:
            npos = pos[0] + d[0], pos[1] + d[1]
            if npos in revealed:
                npos = npos[0] + d[0], npos[1] + d[1]
                if npos not in checked:
                    to_check.append((npos, dist + 1))
    return distances


data = get_data()[0]
revealed = generate_map(data)
distances = flood(revealed)

first = max(distances)
second = sum(1 for i in distances if i >= 1000)

print("First:  ", first)
print("Second: ", second)
